torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -193,49 +220,57 @@ FusionExecutorCache::FusionExecutorCache(
   has_reduction_ = fusion_->hasReduction();
 }
 
-// TODO: dummy cache
 std::vector<at::Tensor> FusionExecutorCache::runFusionWithInputs(
-    const at::ArrayRef<IValue>& inputs) {
-  // caching strategy is different for pw-fusion and reduction-fusion.
-  if (has_reduction_) {
-    // copy the fusion, since each FusionExecutor needs to manipulate the fusion
-    // in order to generate kernel.
-    Fusion fusion = *fusion_;
-    FusionGuard fg(&fusion);
-    TensorView* red_tv = nullptr;
-    for (auto expr : fusion.exprs()) {
-      if (expr->getExprType().has_value() &&
-          expr->getExprType().value() == ExprType::ReductionOp) {
-        red_tv = expr->outputs()[0]->as<TensorView>();
-        break;
+    const at::ArrayRef<IValue>& inputs,
+    size_t unique_id) {
+  if (code_to_fe_lookup_.count(unique_id) == 0) {
+    // enter when we get a new input set. We need to search for compatible
+    // entries in cached `FusionExecutor` or compile new one as needed.
+
+    // caching strategy is different for pw-fusion and reduction-fusion.
+    if (has_reduction_) {
+      // copy the fusion, since each FusionExecutor needs to manipulate the
+      // fusion in order to generate kernel.
+      Fusion fusion = *fusion_;
```

Comment:
We have to copy the fusion on every reduction run?!

Reply:
I think that's the case, since we did basically call `auto reduction_params = scheduleReduction(&fusion, inputs, red_tv);` to get reduction_params.
@kevinstephano in case I'm just using it wrong.